inc/Engine/Optimization/AssetsLocalCache.php

```diff
@@ -0,0 +1,103 @@
+<?php
+namespace WP_Rocket\Engine\Optimization;
+
+use WP_Filesystem_Direct;
+
+/**
+ * Cache locally 3rd party assets.
+ *
+ * @since 3.1
+ */
+class AssetsLocalCache {
+	/**
+	 * 3rd party assets cache folder path
+	 *
+	 * @since 3.1
+	 *
+	 * @var string
+	 */
+	protected $cache_path;
+
+	/**
+	 * Filesystem instance.
+	 *
+	 * @var WP_Filesystem_Direct
+	 */
+	private $filesystem;
+
+	/**
+	 * Constructor
+	 *
+	 * @since 3.1
+	 *
+	 * @param string               $cache_path 3rd party assets cache folder path.
+	 * @param WP_Filesystem_Direct $filesystem Filesysten instance.
+	 */
+	public function __construct( $cache_path, $filesystem ) {
+		$this->cache_path = "{$cache_path}3rd-party/";
+		$this->filesystem = $filesystem;
+	}
+
+	/**
+	 * Gets content for the provided URL.
+	 * Use the local cache file if it exists, else get it from the 3rd party URL and save it locally for future use.
+	 *
+	 * @since 3.1
+	 *
+	 * @param string $url URL to get the content from.
+	 * @return string
+	 */
+	public function get_content( $url ) {
+		$filepath = $this->get_filepath( $url );
+
+		if ( $this->filesystem->is_readable( $filepath ) ) {
+			return $this->filesystem->get_contents( $filepath );
+		}
+
+		$content = wp_remote_retrieve_body( wp_remote_get( $url ) );
+
+		if ( empty( $content ) ) {
+			return '';
+		}
+
+		$this->write_file( $content, $filepath );
+
+		return $content;
+	}
+
+	/**
+	 * Gets the filepath of the local copy for the given URL
+	 *
+	 * @since 3.7
+	 *
+	 * @param string $url URL to get filepath for.
+	 * @return string
+	 */
+	protected function get_filepath( $url ) {
+		$parts    = wp_parse_url( $url );
+		$filename = $parts['host'] . str_replace( '/', '-', $parts['path'] );
```

Comment:
Is it possible that we could get here and there's no path, i.e. `$parts['path']` is empty?

Reply:
It's unlikely, and I don't think we ever had this reported before, but I've seen more unlikely things happening, so I added a guard clause: https://github.com/wp-media/wp-rocket/pull/2785/commits/812b7ae185b1ab4aa2a825edfeea4d449123e3ce